src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -207,10 +209,78 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
                 }
             }
 
+            // We may get a flood of diagnostic notificaitons on the BG.  In order to not overwhelm
+            // the UI thread with work to do, and to avoid excess allocations, we batch up these
+            // notifications, and attempt to process them all at once on the UI thread every 50ms
+            // or so.
+            //
+            // Because we're batching, we can optimize things such that we only store two pieces
+            // of data per provider.  Specifically, if they were removing all diagnostics, and
+            // per-document, the latest diagnostics we've created for it.
+            private static readonly ObjectPool<ProviderAndDocumentToBatchedUpdates> s_providerPool = new ObjectPool<ProviderAndDocumentToBatchedUpdates>(
+                () => new ProviderAndDocumentToBatchedUpdates());
+
+            private readonly object _gate = new object();
+            private ProviderAndDocumentToBatchedUpdates _idToBatchedUpdates = s_providerPool.Allocate();
+            private Task _updateTask = null;
+
             private void OnDiagnosticsUpdatedOnBackground(DiagnosticsUpdatedArgs e)
             {
                 this.AssertIsBackground();
-                RegisterNotification(() => OnDiagnosticsUpdatedOnForeground(e));
+                if (_disposed)
```

Comment:
shouldn't this under lock?

Reply:
No need for lock here.  At worst, we miss some write of 'true' to disposed.  But then the UI request that executes will immediately bail.

Reply:
Note: even under a lock we could 'still' miss things, as this field is always racey if accessed from different threads.  The only important thing is that UI calls respect it and don't do work if it is set to true.  the BG just uses it when it can to avoid unnecessary work, but it's ok to miss.